src/session.ts

```diff
@@ -1,15 +1,14 @@
 import { Context } from './context'
-import { MaybePromise } from './composer'
 import { MiddlewareFn } from './middleware'
 
 export interface SessionStore<T> {
-  get: (name: string) => MaybePromise<T | undefined>
-  set: (name: string, value: T) => MaybePromise<void>
-  delete: (name: string) => MaybePromise<void>
+  readonly get: (name: string) => T | undefined
+  readonly set: (name: string, value: T) => void
+  readonly delete: (name: string) => void
```

Comment:
What kind of session do you expect not to have async operations? Doesn't that completely defeat the point of persistence?

Reply:
Mutable object is a bad database interface. So, storage must store all data in memory, but it can still persist it.

Reply:
But how to persist something synchronously? I assume you aren't suggesting to just dump some floating promises?

Reply:
That, or sync IO, or `setInterval`...

Reply:
Do you believe that this is a good API?

Reply:
No. Good API needs transactions. That's the best I can do without them.

Reply:
I vote for either introducing that transaction support, or dropping session altogether, rather than adding a module which we agree is bad.

Reply:
API would be completely different. If I do it, it'll be a separate package, independent from Telegraf.

Reply:
I.e. drop sessions?

Reply:
That's more drastic than restricting it to be safe, but I could still do that.

Reply:
That is true, but it would also be the more consequent step. I leave it up to you

Reply:
Nothing stops users from using async 3rd party session.

Reply:
Correct, then feel free to go ahead with removing it from core

Reply:
We won't remove it from core - I need more time to think about this. After working on the other stuff pending.

Reply:
> Nothing stops users from using async 3rd party session.

But since it's an interface and plugins will probably use it, won't these 3rd party sessions be incompatible with plugins?